Add random tests with match_only_text multi-field

parkertimmins opened the pull request:
Add tests to the logsdb randomized tests which use multi-fields for each of the `text`, `keyword`, `match_only_text`, and `wildcard` types. For each type allow each other type of multi-field.

parkertimmins committed:
Add random tests for match_only_text with multi-field

parkertimmins committed:
Skip phrase query for match_only_text in nested

parkertimmins committed:
Simplify phrase query generation

parkertimmins committed:
Tune down number of ignove_above and multi-fields

parkertimmins committed:
Some cleanup and tune params

Commit:
[CI] Auto commit changes from spotless

parkertimmins committed:
Use text matcher for match_only_text

parkertimmins commented:
The updates in the PR should have caught the bug fixed in https://github.com/elastic/elasticsearch/pull/131314 and https://github.com/elastic/elasticsearch/pull/131383. To verify that this is the case, used the patch below which reverts the relevant code to the pre-fix state.

Running: 
`./gradlew :x-pack:plugin:logsdb:javaRestTest --tests "org.elasticsearch.xpack.logsdb.qa.BulkChallengeRestIT.testRandomQueries" -Dtests.iters=100` with the old code from the patch usually results in a few failures. Some with the expected messages, like ` Cannot invoke "java.util.List.iterator()" because "values" is null`


```
Index: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java b/server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java
--- a/server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java	(revision 560d937fb913742ebfa548f6fbad0f6cae93d0a7)
+++ b/server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java	(date 1754276716082)
@@ -547,7 +547,6 @@
         private final IndexMode indexMode;
         private final IndexSortConfig indexSortConfig;
         private final boolean hasDocValuesSkipper;
-        private final String originalName;
 
         public KeywordFieldType(
             String name,
@@ -576,7 +575,6 @@
             this.indexMode = builder.indexMode;
             this.indexSortConfig = builder.indexSortConfig;
             this.hasDocValuesSkipper = DocValuesSkipIndexType.NONE.equals(fieldType.docValuesSkipIndexType()) == false;
-            this.originalName = isSyntheticSource ? name + "._original" : null;
         }
 
         public KeywordFieldType(String name, boolean isIndexed, boolean hasDocValues, Map<String, String> meta) {
@@ -591,7 +589,6 @@
             this.indexMode = IndexMode.STANDARD;
             this.indexSortConfig = null;
             this.hasDocValuesSkipper = false;
-            this.originalName = null;
         }
 
         public KeywordFieldType(String name) {
@@ -617,7 +614,6 @@
             this.indexMode = IndexMode.STANDARD;
             this.indexSortConfig = null;
             this.hasDocValuesSkipper = DocValuesSkipIndexType.NONE.equals(fieldType.docValuesSkipIndexType()) == false;
-            this.originalName = null;
         }
 
         public KeywordFieldType(String name, NamedAnalyzer analyzer) {
@@ -632,7 +628,6 @@
             this.indexMode = IndexMode.STANDARD;
             this.indexSortConfig = null;
             this.hasDocValuesSkipper = false;
-            this.originalName = null;
         }
 
         @Override
@@ -1096,15 +1091,6 @@
         ) {
             return new AutomatonQueryWithDescription(new Term(name()), automatonSupplier.get(), description);
         }
-
-        /**
-         * The name used to store "original" that have been ignored
-         * by {@link KeywordFieldType#ignoreAbove()} so that they can be rebuilt
-         * for synthetic source.
-         */
-        public String originalName() {
-            return originalName;
-        }
     }
 
     private final boolean indexed;
@@ -1158,7 +1144,7 @@
         this.forceDocValuesSkipper = builder.forceDocValuesSkipper;
         this.offsetsFieldName = offsetsFieldName;
         this.indexSourceKeepMode = indexSourceKeepMode;
-        this.originalName = mappedFieldType.originalName();
+        this.originalName = isSyntheticSource ? fullPath() + "._original" : null;
     }
 
     @Override
@@ -1218,7 +1204,7 @@
                 // Save a copy of the field so synthetic source can load it
                 var utfBytes = value.bytes();
                 var bytesRef = new BytesRef(utfBytes.bytes(), utfBytes.offset(), utfBytes.length());
-                context.doc().add(new StoredField(originalName, bytesRef));
+                context.doc().add(new StoredField(originalName(), bytesRef));
             }
             return false;
         }
@@ -1334,6 +1320,15 @@
         return normalizerName != null;
     }
 
+    /**
+     * The name used to store "original" that have been ignored
+     * by {@link KeywordFieldType#ignoreAbove()} so that they can be rebuilt
+     * for synthetic source.
+     */
+    private String originalName() {
+        return originalName;
+    }
+
     @Override
     protected SyntheticSourceSupport syntheticSourceSupport() {
         if (hasNormalizer()) {
@@ -1382,7 +1377,7 @@
         }
 
         if (fieldType().ignoreAbove != Integer.MAX_VALUE) {
-            layers.add(new CompositeSyntheticFieldLoader.StoredFieldLayer(originalName) {
+            layers.add(new CompositeSyntheticFieldLoader.StoredFieldLayer(originalName()) {
                 @Override
                 protected void writeValue(Object value, XContentBuilder b) throws IOException {
                     BytesRef ref = (BytesRef) value;
Index: modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java b/modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java
--- a/modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java	(revision 560d937fb913742ebfa548f6fbad0f6cae93d0a7)
+++ b/modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java	(date 1754276716081)
@@ -47,7 +47,6 @@
 import org.elasticsearch.index.mapper.BlockStoredFieldsReader;
 import org.elasticsearch.index.mapper.DocumentParserContext;
 import org.elasticsearch.index.mapper.FieldMapper;
-import org.elasticsearch.index.mapper.KeywordFieldMapper;
 import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.index.mapper.MapperBuilderContext;
 import org.elasticsearch.index.mapper.SourceValueFetcher;
@@ -252,17 +251,6 @@
             if (searchExecutionContext.isSourceSynthetic() && withinMultiField) {
                 String parentField = searchExecutionContext.parentPath(name());
                 var parent = searchExecutionContext.lookup().fieldType(parentField);
-
-                if (parent instanceof KeywordFieldMapper.KeywordFieldType keywordParent
-                    && keywordParent.ignoreAbove() != Integer.MAX_VALUE) {
-                    if (parent.isStored()) {
-                        return storedFieldFetcher(parentField, keywordParent.originalName());
-                    } else if (parent.hasDocValues()) {
-                        var ifd = searchExecutionContext.getForField(parent, MappedFieldType.FielddataOperation.SEARCH);
-                        return combineFieldFetchers(docValuesFieldFetcher(ifd), storedFieldFetcher(keywordParent.originalName()));
-                    }
-                }
-
                 if (parent.isStored()) {
                     return storedFieldFetcher(parentField);
                 } else if (parent.hasDocValues()) {
@@ -274,19 +262,8 @@
             } else if (searchExecutionContext.isSourceSynthetic() && hasCompatibleMultiFields) {
                 var mapper = (MatchOnlyTextFieldMapper) searchExecutionContext.getMappingLookup().getMapper(name());
                 var kwd = TextFieldMapper.SyntheticSourceHelper.getKeywordFieldMapperForSyntheticSource(mapper);
-
                 if (kwd != null) {
                     var fieldType = kwd.fieldType();
-
-                    if (fieldType.ignoreAbove() != Integer.MAX_VALUE) {
-                        if (fieldType.isStored()) {
-                            return storedFieldFetcher(fieldType.name(), fieldType.originalName());
-                        } else if (fieldType.hasDocValues()) {
-                            var ifd = searchExecutionContext.getForField(fieldType, MappedFieldType.FielddataOperation.SEARCH);
-                            return combineFieldFetchers(docValuesFieldFetcher(ifd), storedFieldFetcher(fieldType.originalName()));
-                        }
-                    }
-
                     if (fieldType.isStored()) {
                         return storedFieldFetcher(fieldType.name());
                     } else if (fieldType.hasDocValues()) {
@@ -335,52 +312,13 @@
             };
         }
 
-        private static IOFunction<LeafReaderContext, CheckedIntFunction<List<Object>, IOException>> storedFieldFetcher(String... names) {
-            var loader = StoredFieldLoader.create(false, Set.of(names));
+        private static IOFunction<LeafReaderContext, CheckedIntFunction<List<Object>, IOException>> storedFieldFetcher(String name) {
+            var loader = StoredFieldLoader.create(false, Set.of(name));
             return context -> {
                 var leafLoader = loader.getLoader(context, null);
                 return docId -> {
                     leafLoader.advanceTo(docId);
-                    var storedFields = leafLoader.storedFields();
-                    if (names.length == 1) {
-                        return storedFields.get(names[0]);
-                    }
-
-                    List<Object> values = new ArrayList<>();
-                    for (var name : names) {
-                        var currValues = storedFields.get(name);
-                        if (currValues != null) {
-                            values.addAll(currValues);
-                        }
-                    }
-
-                    return values;
-                };
-            };
-        }
-
-        private static IOFunction<LeafReaderContext, CheckedIntFunction<List<Object>, IOException>> combineFieldFetchers(
-            IOFunction<LeafReaderContext, CheckedIntFunction<List<Object>, IOException>> primaryFetcher,
-            IOFunction<LeafReaderContext, CheckedIntFunction<List<Object>, IOException>> secondaryFetcher
-        ) {
-            return context -> {
-                var primaryGetter = primaryFetcher.apply(context);
-                var secondaryGetter = secondaryFetcher.apply(context);
-                return docId -> {
-                    List<Object> values = new ArrayList<>();
-                    var primary = primaryGetter.apply(docId);
-                    if (primary != null) {
-                        values.addAll(primary);
-                    }
-
-                    var secondary = secondaryGetter.apply(docId);
-                    if (secondary != null) {
-                        values.addAll(secondary);
-                    }
-
-                    assert primary != null || secondary != null;
-
-                    return values;
+                    return leafLoader.storedFields().get(name);
                 };
             };
         }

```

parkertimmins committed:
Merge branch 'main' into parker/random-tests-match-only-text-multifield

martijnvg reviewed (approved):
Nice work - LGTM 👍 

I think the `>test` label is more appropriate here?

parkertimmins committed:
Merge branch 'main' into parker/random-tests-match-only-text-multifield

parkertimmins committed:
Remove wildcard as a subfield option

Ideally, wildcard would be tested as a subfield. But data generation
code is used in server, and wildcard type is in xpack. With some better
wiring this could be fixed, but that will have to wait for the future.

parkertimmins committed:
Merge branch 'main' into parker/random-tests-match-only-text-multifield

parkertimmins committed:
Dont add wildcard and match_only_text to block loader tests

parkertimmins committed:
Merge branch 'main' into parker/random-tests-match-only-text-multifield

parkertimmins committed:
Update multi-field name to subfield_keyword

lkts reviewed (commented):
I have a suggestion to handle `includePluginTypesInMultiFields` in a cleaner way.

parkertimmins committed:
Move multifield into separate data source handler

Commit:
[CI] Auto commit changes from spotless

parkertimmins committed:
Small fix

parkertimmins committed:
Change set to list to avoid non-determinism

parkertimmins committed:
improve naming

lkts commented on test/framework/src/main/java/org/elasticsearch/datageneration/DataGeneratorSpecification.java:
```diff
@@ -26,14 +26,16 @@
     boolean fullyDynamicMapping,
-    List<PredefinedField> predefinedFields
+    List<PredefinedField> predefinedFields,
+    boolean includePluginTypes
```
Yes, if you can implement it in the "main" module it's fine. `geo_shape` is special because it relies on test helpers defined inside geo module.

parkertimmins commented on test/framework/src/main/java/org/elasticsearch/datageneration/datasource/MultifieldAddonHandler.java:
```diff
@@ -0,0 +1,99 @@
+    private static Map<String, Object> getChildMappingForType(FieldType type, DataSourceRequest.LeafMappingParametersGenerator request) {
+        Map<String, Object> mapping = getMappingForType(type, request);
+        if (type == FieldType.KEYWORD) {
+            mapping.remove("copy_to");
```
Oh yeah, I guess it should probably be done for other types too :thinking:

parkertimmins committed:
Improvements from review

parkertimmins committed:
Merge branch 'main' into parker/random-tests-match-only-text-multifield
